cluster-autoscaler/proposals/buffers.md

```diff
@@ -0,0 +1,567 @@
+    monitoring and insights into usage and needs in the cluster.
+
+There was already a similar proposal that was not implemented in the end:
+[Pod headroom](https://docs.google.com/document/d/1SyqStWUt407Rcwdtv25yG6MpHdNnbfB3KPmc4zQuz1M/edit?pli=1&tab=t.0#heading=h.coxzr2ev5x0k)
+
+The feature has been requested multiple times by the community
+([#749](https://github.com/kubernetes-sigs/karpenter/issues/749),
+[#987](https://github.com/aws/karpenter-provider-aws/issues/987),
+[#3240](https://github.com/aws/karpenter-provider-aws/issues/3240),
+[#3384](https://github.com/aws/karpenter-provider-aws/issues/3384),
+[#4409](https://github.com/aws/karpenter-provider-aws/issues/4409))
+
+## Goals
+
+*   User can define spare capacity in the cluster that should be available for
+    the user workloads
+*   User pods can use the available capacity without any configuration changes
+    (strictly additive, can be managed by the user who owns the pod as well as
+    cluster administrators)
+*   Open possibility for cloud providers to offer different types of buffers if
+    needed
+
+## Non-Goals
+
+*   In the initial implementation the buffers will not be guaranteed/reserved
+    for any particular pods, any pod that can run on a given capacity will be
+    able to use it. Once scheduler offers concept of reserving cluster capacity
+    it will be possible to add configuration options with hard guarantees (for
+    example when a user wants to keep a buffer only for top priority pods, but
+    on the same nodes there are also other pods running)
+*   Providing spare capacity depends heavily on the autoscaling capabilities of
+    the cluster - the support for buffers will depend on the compatibility of
+    the autoscaler used within the cluster.
+
+# Proposal
+
+A `Buffer` CRD (autoscaling.x-k8s.io) is added to represent spare capacity
+requested by the user along with a set of libraries that make it easy to
+integrate any autoscaler along with reference implementation in the
+[Cluster Autoscaler repository](https://github.com/kubernetes/autoscaler).
+
+In order to support buffers the cluster will need to run:
+
+*   A controller that translates the buffer configuration into a number of pods
+    specs that would represent the spare space in the cluster that is needed.
+    The reference implementation would be provided within the cluster autoscaler
+    repository, but it is expected that most integrations will depend on it as a
+    library rather than full implementation in order to support
+    autoscaler/provider specific configurations like CCC (in GKE) or NodePool
+    (in karpenter)
+*   An autoscaler that besides processing real pods should allow for processing
+    additional pods that are representing a buffer (for example treating them as
+    virtual pods) so that the capacity in the cluster becomes available.
+
+### How do we determine the pod spec for targets using scale subresource?
```

Comment:
Why not use a pod template spec inline? Imagine the case where you want to leave room for a CI job, but the job doesn't exist in the cluster yet, so you have nothing to reference.